packages/field-plugin/src/createFieldPlugin/createPluginActions/callbackQueue.ts

```diff
@@ -0,0 +1,66 @@
+import {
+  AssetSelectedMessage,
+  ContextRequestMessage,
+  OnMessage,
+  LoadedMessage,
+} from '../../messaging'
+
+type CallbackId = string
+type CallbackRef<Message> = {
+  callbackId: CallbackId
+  callback: OnMessage<Message>
+}
+type CallbackMap = {
+  asset: CallbackRef<AssetSelectedMessage>[]
+  context: CallbackRef<ContextRequestMessage>[]
+  state: CallbackRef<LoadedMessage>[]
+}
+type CallbackType = keyof CallbackMap
+
+export const callbackQueue = () => {
+  // eslint-disable-next-line functional/no-let
+  let callbackMap: CallbackMap = {
+    asset: [],
+    context: [],
+    state: [],
+  }
+  // eslint-disable-next-line functional/no-let
+  let uuidIndex = 0
+  // TODO randomly generate instead
+  const uuid = (): CallbackId => {
+    uuidIndex = uuidIndex + 1
+    return uuidIndex
+  }
+
+  const pushCallback = <T extends CallbackType>(
+    callbackType: T,
+    callback: CallbackMap[T][number]['callback'],
+  ): CallbackId => {
+    // TODO limit length of array
+    const callbackId = uuid()
+    callbackMap = {
+      ...callbackMap,
+      [callbackType]: [
+        ...callbackMap[callbackType],
+        {
+          callbackId,
+          callback,
+        },
+      ],
+    }
+    return callbackId
+  }
+  const popCallback = <T extends CallbackType>(
+    callbackType: T,
+    callbackId: CallbackId | undefined,
+  ): CallbackMap[T][number]['callback'] | undefined => {
+    // TODO remove callback when popping
+    return callbackMap[callbackType].findLast(
```

Comment:
I'm not sure but don't we need to find the first one?

Reply:
It depends on whether we push items to the front or to the back of the queue. Since each `callbackId` is unique, it will find the element regardless of how it traverses the list, but it's slightly better to check from the end of the array, since the array is sorted by order of insertion.